airflow/utils/db.py

```diff
@@ -1043,103 +1061,106 @@ def _offline_migration(migration_func: Callable, config, revision):
+                f"which corresponds to Airflow {min_version}."
             )
 
 
 @provide_session
 def upgradedb(
-    version_range: Optional[str] = None, revision_range: Optional[str] = None, session: Session = NEW_SESSION
+    to_revision: Optional[str] = None,
+    from_revision: Optional[str] = None,
+    sql: bool = False,
+    session: Session = NEW_SESSION,
 ):
-    """Upgrade the database."""
+    """
+
+    :param to_revision: Optional Alembic revision ID to upgrade *to*.
+        If omitted, upgrades to latest revision.
+    :param from_revision: Optional Alembic revision ID to upgrade *from*.
+        Not compatible with ``sql_only=False``.
+    :param sql: if True, migration statements will be printed but not executed.
+    :param session: sqlalchemy session with connection to Airflow metadata database
+    :return: None
+    :rtype: None
+    """
+    if from_revision and not sql:
+        raise Exception("`from_revision` only supported with `sql_only=True`.")
+
     # alembic adds significant import time, so we import it lazily
     if not settings.SQL_ALCHEMY_CONN:
-        raise RuntimeError("The settings.SQL_ALCHEMY_CONN not set. This is critical assertion.")
+        raise RuntimeError("The settings.SQL_ALCHEMY_CONN not set. This is a critical assertion.")
     from alembic import command
-    from alembic.script import ScriptDirectory
 
     config = _get_alembic_config()
-    script_ = ScriptDirectory.from_config(config)
 
-    config.set_main_option('sqlalchemy.url', settings.SQL_ALCHEMY_CONN.replace('%', '%%'))
-    if version_range:
-        revision = _validate_version_range(script_, version_range)
-        if not revision:
+    if sql:
+        if not from_revision:
+            from_revision = _get_current_revision(session)
+
+        if not from_revision and to_revision:
+            raise Exception('unexpected')
+
+        if to_revision == from_revision:
+            print_happy_cat("No migrations to apply; nothing to do.")
             return
-        log.info("Running offline migrations for version range %s", version_range)
-        return _offline_migration(command.upgrade, config, revision)
-    elif revision_range:
-        _validate_revision(script_, revision_range)
-        log.info("Running offline migrations for revision range %s", revision_range)
-        return _offline_migration(command.upgrade, config, revision_range)
+
+        if not _revision_greater(config, to_revision, from_revision):
+            raise ValueError(
+                f'Requested *to* revision {to_revision} is older than *from* revision {from_revision}. '
+                'Please check your requested versions / revisions.'
+            )
+        _revisions_above_min_for_offline(config=config, revisions=[from_revision, to_revision])
+
+        _offline_migration(command.upgrade, config, f"{from_revision}:{to_revision}")
+        return
```

Comment:
```suggestion
```

Reply:
The return is necessary to avoid it continuing to run the real upgrade command. Maybe we should return the _offline_migration?

Reply:
> The return is necessary to avoid it continuing to run the real upgrade command. Maybe we should return the _offline_migration?

i liked the empty return because returning the func call makes the reader think "hmm is there something returned from that func that we aree returning again here?" but empty return (or return None) makes it obvious that _nothing is returned_ 

i can add a comment on the line to explain the intention that tripped jed up

Reply:
Thanks, I'm going to blame the github diff, not my tired eyes last night 🙃